cyral/internal/policy/v2/model.go

```diff
@@ -14,127 +21,200 @@ type ChangeInfo struct {
 	}
-	if err := d.Set("valid_until", r.Policy.ValidUntil); err != nil {
+	if err := d.Set("valid_until", timestampFromProtobuf(p.GetValidUntil())); err != nil {
 		return fmt.Errorf("error setting 'valid_until' field: %w", err)
 	}
-	if err := d.Set("document", r.Policy.Document); err != nil {
+	if err := d.Set("document", p.GetDocument()); err != nil {
 		return fmt.Errorf("error setting 'document' field: %w", err)
 	}
-
-	// Use the ToMap method to set the last_updated and created fields
-	if err := d.Set("last_updated", r.Policy.LastUpdated.ToMap()); err != nil {
+	// Use the changeInfoToMap method to set the last_updated and created fields
+	if err := d.Set("last_updated", changeInfoToMap(p.GetLastUpdated())); err != nil {
 		return fmt.Errorf("error setting 'last_updated' field: %w", err)
 	}
-	if err := d.Set("created", r.Policy.Created.ToMap()); err != nil {
+	if err := d.Set("created", changeInfoToMap(p.GetCreated())); err != nil {
 		return fmt.Errorf("error setting 'created' field: %w", err)
 	}
-	if err := d.Set("enforced", r.Policy.Enforced); err != nil {
+	if err := d.Set("enforced", p.GetEnforced()); err != nil {
 		return fmt.Errorf("error setting 'enforced' field: %w", err)
 	}
-	if r.Policy.Type != "" {
-		if err := d.Set("type", r.Policy.Type); err != nil {
+	// policy types have aliases, so we don't want to set the policy type
+	// except if the new value is not an alias for the old one.
+	if msg.PolicyType_value[d.Get("type").(string)] != int32(ptype) {
+		if err := d.Set("type", ptype.String()); err != nil {
 			return fmt.Errorf("error setting 'type' field: %w", err)
 		}
 	}
-	if r.Policy.Scope != nil {
-		if err := d.Set("scope", r.Policy.Scope.ToMap()); err != nil {
+	if p.GetScope() != nil {
+		if err := d.Set("scope", scopeToMap(p.GetScope())); err != nil {
 			return fmt.Errorf("error setting 'scope' field: %w", err)
 		}
 	}
-	d.SetId(r.Policy.ID)
+	d.SetId(p.GetId())
 	return nil
 }
 
-// ReadFromSchema reads the policy data from the schema
-func (r *PolicyV2) ReadFromSchema(d *schema.ResourceData) error {
-	r.Policy.ID = d.Get("id").(string)
-	r.Policy.Name = d.Get("name").(string)
-	r.Policy.Description = d.Get("description").(string)
-	r.Policy.Enabled = d.Get("enabled").(bool)
-	r.Policy.Tags = utils.ConvertFromInterfaceList[string](d.Get("tags").([]interface{}))
-	r.Policy.ValidFrom = d.Get("valid_from").(string)
-	r.Policy.ValidUntil = d.Get("valid_until").(string)
-	r.Policy.Document = d.Get("document").(string)
-	r.Policy.Enforced = d.Get("enforced").(bool)
-	r.Policy.Type = d.Get("type").(string)
+func timestampFromResourceData(key string, d *schema.ResourceData) (*timestamppb.Timestamp, error) {
+	ts := d.Get(key).(string)
+	if ts == "" {
+		return nil, nil
+	}
+	t, err := time.Parse(time.RFC3339, ts)
+	if err != nil {
+		return nil, fmt.Errorf("invalid %s value: %s", key, ts)
+	}
+	return timestamppb.New(t), nil
+}
+
+func timestampFromProtobuf(ts *timestamppb.Timestamp) string {
```

Comment:
These timestamp helper functions I believe we will need for other migrated resources if you think it is worthwhile to collect some of these common helpers.

Reply:
Done